File: src/data/courses/dsa/trie/counting-dictionary/sum-prefix-scores/code/java/SumPrefixScores.java
// ============================================================
// Solution 1: Brute Force — HashSet / Direct comparison
// Time: O(N*L) or O(N²*L) | Space: O(N*L)
// ============================================================
import java.util.*;
class Solution1 {
    Set<String> words = new HashSet<>();
    public void insert(String word) { words.add(word); }
    public boolean hasPrefix(String prefix) {
        for (String w : words)
            if (w.startsWith(prefix)) return true;
        return false;
    }
}
// ============================================================
// Solution 2: Optimal (Trie-based)
// ============================================================
: Optimal (Trie-based)
// ============================================================
class SumPrefixScores {
    class TrieNode {
        TrieNode[] children = new TrieNode[26];
        int count = 0; // Count of words passing through
    }
    
    private TrieNode root = new TrieNode();
    
    public int[] sumPrefixScores(String[] words) {
        // Insert all words
        for (String word : words) insert(word);
        
        // Calculate scores
        int[] result = new int[words.length];
        for (int i = 0; i < words.length; i++) {
            result[i] = getScore(words[i]);
        }
        return result;
    }
    
    private void insert(String word) {
        TrieNode node = root;
        for (char c : word.toCharArray()) {
            int idx = c - 'a';
            if (node.children[idx] == null) {
                node.children[idx] = new TrieNode();
            }
            node = node.children[idx];
            node.count++; // Increment count
        }
    }
    
    private int getScore(String word) {
        TrieNode node = root;
        int score = 0;
        for (char c : word.toCharArray()) {
            int idx = c - 'a';
            node = node.children[idx];
            score += node.count; // Sum all prefix counts
        }
        return score;
    }
}

File: src/data/courses/dsa/trie/counting-dictionary/sum-prefix-scores/code/cpp/SumPrefixScores.cpp
// ============================================================
// Solution 1: For each word, count prefix matches in all words
// Time: O(N² * L) | Space: O(1) extra
// ============================================================
#include <vector>
#include <string>
using namespace std;

class Solution1 {
public:
    vector<int> sumPrefixScores(vector<string>& words) {
        int n = words.size();
        vector<int> scores(n, 0);
        for (int i = 0; i < n; i++) {
            for (int len = 1; len <= (int)words[i].size(); len++) {
                string prefix = words[i].substr(0, len);
                for (int j = 0; j < n; j++)
                    if (words[j].size() >= prefix.size() && words[j].substr(0, len) == prefix)
                        scores[i]++;
            }
        }
        return scores;  // O(N²*L) — check each prefix against all words
    }
};

// ============================================================
// Solution 2: Trie with prefix count at each node (Optimal)
// Time: O(N * L) | Space: O(N * L)
// ============================================================
struct TrieNode {
    TrieNode* children[26] = {};
    int count = 0;  // How many words pass through this node
};

class Solution2 {
    TrieNode* root = new TrieNode();
public:
    void insert(const string& word) {
        TrieNode* node = root;
        for (char c : word) {
            if (!node->children[c - 'a'])
                node->children[c - 'a'] = new TrieNode();
            node = node->children[c - 'a'];
            node->count++;  // Increment prefix count
        }
    }

    int getPrefixScore(const string& word) {
        TrieNode* node = root;
        int score = 0;
        for (char c : word) {
            node = node->children[c - 'a'];
            score += node->count;  // Add count at each prefix node
        }
        return score;
    }

    vector<int> sumPrefixScores(vector<string>& words) {
        for (auto& w : words) insert(w);
        vector<int> result;
        for (auto& w : words) result.push_back(getPrefixScore(w));
        return result;
    }
};
